lib/mongoid/atomic.rb

```diff
@@ -385,11 +385,35 @@ def touch_atomic_updates(field = nil)
       updates = atomic_updates
       return {} unless atomic_updates.key?("$set")
       touches = {}
+      wanted_keys = %w(updated_at u_at)
+      # TODO this permits field to be passed as an empty string in which case
+      # it is ignored, get rid of this behavior.
+      if field.present?
+        wanted_keys << field.to_s
+      end
       updates["$set"].each_pair do |key, value|
-        key_regex = /updated_at|u_at#{"|" + field if field.present?}/
-        touches.merge!({ key => value }) if key =~ key_regex
+        if wanted_keys.include?(key.split('.').last)
+          touches.update(key => value)
+        end
       end
       { "$set" => touches }
     end
+
+    # Returns the $set atomic updates affecting the specified field.
+    #
+    # @param [ String ] field The field name.
+    #
+    # @api private
+    def set_field_atomic_updates(field)
+      updates = atomic_updates
+      return {} unless atomic_updates.key?("$set")
+      sets = {}
+      updates["$set"].each_pair do |key, value|
+        if key.split('.').last == field
+          sets.update(key => value)
+        end
+      end
+      { "$set" => sets }
+    end
```

Comment:
Instead of doing this, how about adding `atomic_touches` as another element in `Mongoid::Atomic::UPDATES`. Then you can implement the solution in `MONGOID-5142` when you resolve the atomic actions at the end, but merging it an existing `set` update if one is available.

Reply:
Hi Johnny,

My thinking with this PR and 5142 was to repair the problem using the quicker solution so that, in particular, I could attend to the backlog of PRs (many of which are, as  I see, created by you - thank you for that).

If you'd like to create a follow-up to this PR that addresses 5142 I'll appreciate it.

Reply:
Sure, I'll take a look at this tomorrow.

My pleasure creating the PRs, you help to review them is greatly appreciated.

FYI in addition I'm going to look at:
- https://jira.mongodb.org/projects/MONGOID/issues/MONGOID-5028
- https://jira.mongodb.org/projects/MONGOID/issues/MONGOID-5122
- https://jira.mongodb.org/projects/MONGOID/issues/MONGOID-5022 (maybe)

Reply:
Fix for MONGOID-5142 here: #5045, this PR can be closed.